### PR TITLE
[dcp] Integrate stream extensions into DCP impl

### PR DESCRIPTION
**Pull request by mhorowitz**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #143360
* __->__ #143359
* #143358

Summary: Updates FileSystemReader/Writer, Planner, DefaultLoad/SavePlanner

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @pradeepfn @ekr0

### CODE REVIEW

File: torch/distributed/checkpoint/default_planner.py [hunk 1]
```diff
@@ -72,11 +73,13 @@ def __init__(
         flatten_sharded_tensors: bool = True,
         dedup_replicated_tensors: Optional[bool] = None,
         dedup_save_to_lowest_rank: bool = False,
+        extensions: Optional[Sequence[StreamTransformExtension]] = None,
```
Comment (LucasLLC) [inline torch/distributed/checkpoint/default_planner.py:76]: I had kind of assumed this made more sense in the file writer since it's closely related to the serialization format, and generally we've thought about planners dealing with 'sharding' specific transforms. Am I missing something?

File: torch/distributed/checkpoint/default_planner.py [hunk 2]
```diff
@@ -169,12 +202,16 @@ def __init__(
         flatten_state_dict: bool = True,
         flatten_sharded_tensors: bool = True,
         allow_partial_load: bool = False,
+        extension_registry: Optional[ExtensionRegistry] = None,
```
Comment (LucasLLC) [inline torch/distributed/checkpoint/default_planner.py:205]: Just flagging one of the largest complaints from our users is that the planner/writer abstraction is too complicated and hard to understand what's going on. I wonder if the planner is the right place for this kind of transformation and whether we've explored doing this somewhere else?

Comment (mhorowitz) [reply torch/distributed/checkpoint/default_planner.py:205]: I think it's hard to understand because there aren't clear delineations of responsibility.  See the previous comment about serialization not being in a single place.  This is absolutely an area where DCP can improve.  I tried doing some cleanup but ran into some difficulties, so I shelved those cleanups in favor of solving the specific compression problem to start.

More generally, though, it seems valuable to reuse common code.  For both serialization and stream transforms, it seems preferable to provide a common implementation which all reader/writers can use.  Otherwise, improvements in these areas need to be made once per implementation, which is made worse by the possibility of local implementations which we don't control.  By using planner hooks which the reader/writer can call into, we give ourselves code sharing and flexibility.  That's why the code is structured the way it is.  We don't want to add compression separately to each reader/writer.

The pattern we should (I believe) encourage is for the reader/writer to specify the storage mechanism (filesystem, fsspec, S3, manifold, etc), and for the implementation to call into the planner for common logic (sharding, serialization, compression, batching, etc).  This makes adding new storage mechanisms easier, and means many improvements can be made centrally, but still gives flexibility to override the common patterns should a storage developer want to (I'm not sure why they would).

That said, you've correctly identified one of the big directional areas in this stack, and I'm happy to discuss it further if you think this is the wrong tradeoff.

File: torch/distributed/checkpoint/default_planner.py (hunk 1 above)
Comment (saumishr) [reply torch/distributed/checkpoint/default_planner.py:76]: Implementing the post processing as a buffer stream transformation makes it out of place in the Planner. I agree on that. Streaming, buffer reading/writing should all be internal to the storage IO. Planner should have no reason to be exposed to the internal mechanics of the storage components. 

However I see value in keeping Optimizer as another component which can be plugged into DCP. This gives the flexibility to the users in bringing the proprietary algorithms or at times our internal algorithms may not be open source, necessitating different versions of the optimizer hooks. 

Open question is where should these hooks be integrated in. If we take the approach of extending the buffer streams then it should be called in the storage components. I was more inclined to use the resolve hooks to do the post processing: sharding and format optimizations both. I prefer to keep the storage components as lean as possible, focusing only on the IO, serialization and deserialization only. Storage components are more likely to get extended/reimplemented by the partners as opposed to the planners. With IO, efficiency optimizations, storage resiliency optimizations, serialization/deserialization and also complex optimizations like compression, batching etc. this will grow significantly in complexity.
Planner should just get the data read from storage and then in resolve hooks, it should do whatever transformations are needed. 

Basically, the way I was thinking about this is that these storage optimizations should neither be in Planner nor in Storage reader/writer IMO but should be somewhat decoupled and invoked as general hooks by the planner at the resolve time.

Comment (saumishr) [reply torch/distributed/checkpoint/default_planner.py]: We discussed offline. This PR is being unblocked to experiment. It won't be a public config yet. We will follow up with a design discussion on this and update the implementation and interfaces accordingly.

File: test/distributed/checkpoint/test_file_system_checkpoint_cpu.py [hunk 3]
```diff
@@ -129,6 +130,40 @@ def test_read_write_only_tensor(self, thread_count) -> None:
...
+class TestDistributedStateDictSaveLoadRot13(TestCase):
+    @parametrize("thread_count", _THREAD_COUNTS)
+    def test_read_write_only_tensor(self, thread_count) -> None:
+        with tempfile.TemporaryDirectory() as path:
+            state_dict_to_save = MyTestModule().state_dict()
+
+            fs_writer = FileSystemWriter(
+                path=path,
+                thread_count=thread_count,
+                _extensions=[Rot13Example()],
```
Comment (saumishr) [inline test/distributed/checkpoint/test_file_system_checkpoint_cpu.py:142]: Why don't we just use Zstandard() extension for tests as well. I don't see a value of having this extra example extension. Seems unnecessary. We are mocking the data already but can test it with an actual extension introduced in the next PR.

Comment (mhorowitz) [reply test/distributed/checkpoint/test_file_system_checkpoint_cpu.py:142]: When I was doing the work on this diff, having a very simple extension was helpful to understand the behavior and debug.  That will continue to be the case as I continue working on this.  Zstandard is a complicated beast, and it's useful to have unit tests which don't depend on it.

**Review by saumishr — APPROVED** [review]
